File: internal/delivery/grpc/handler/ip.go
package handler

import (
	"context"
	"github.com/magomedcoder/ipmanager/api/pb"
	"github.com/magomedcoder/ipmanager/internal/usecase"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"gorm.io/gorm"
)

type IpHandler struct {
	pb.UnimplementedIpServiceServer
	IpUseCase     usecase.IIpUseCase
	SubnetUseCase usecase.ISubnetUseCase
}

func (i *IpHandler) GetIps(ctx context.Context, in *pb.GetIpsRequest) (*pb.GetIpsResponse, error) {
	var total int64
	ips, err := i.IpUseCase.GetIps(ctx, func(db *gorm.DB) {
		db.Where("subnet_id = ?", in.SubnetId).
			Order("id ASC").
			Count(&total)
	})
	if err != nil {
		return nil, status.Error(codes.NotFound, "IP не найдены")
	}

	items := make([]*pb.IpItem, 0)
	for _, item := range ips {
		items = append(items, &pb.IpItem{
			Id:           item.Id,
			Ip:           item.Ip,
			CustomerId:   item.CustomerId,
			CustomerName: item.CustomerName,
			Description:  item.Description,
		})
	}

	return &pb.GetIpsResponse{
		Total: total,
		Items: items,
	}, nil
}

func (i *IpHandler) GetIpById(ctx context.Context, in *pb.GetIpRequest) (*pb.GetIpResponse, error) {
	ip, _ := i.IpUseCase.GetById(ctx, in.Id)
	if ip == nil {
		return nil, status.Error(codes.NotFound, "Ip не найден")
	}

	return &pb.GetIpResponse{
		Id:           ip.Id,
		Ip:           ip.Ip,
		CustomerId:   ip.CustomerId,
		CustomerName: ip.CustomerName,
		Description:  ip.Description,
	}, nil
}

func (i *IpHandler) EditIpCustomer(ctx context.Context, in *pb.EditIpCustomerRequest) (*pb.EditIpCustomerResponse, error) {
	if err := i.IpUseCase.EditCustomerById(ctx, in.Id, in.CustomerId); err != nil {
		return nil, status.Error(codes.NotFound, "Ip не найден")
	}
	return &pb.EditIpCustomerResponse{
		Success: true,
	}, nil
}

func (i *IpHandler) EditIpService(ctx context.Context, in *pb.EditIpServiceRequest) (*pb.EditIpServiceResponse, error) {
	if err := i.IpUseCase.EditServiceById(ctx, in.Id, in.ServiceId); err != nil {
		return nil, status.Error(codes.NotFound, "Ip не найден")
	}
	return &pb.EditIpServiceResponse{
		Success: true,
	}, nil
}

func (i *IpHandler) EditIpDescription(ctx context.Context, in *pb.EditIpDescriptionRequest) (*pb.EditIpDescriptionResponse, error) {
	if err := i.IpUseCase.EditDescriptionById(ctx, in.Id, in.Description); err != nil {
		return nil, status.Error(codes.NotFound, "Ip не найден")
	}

	return &pb.EditIpDescriptionResponse{
		Success: true,
	}, nil
}

File: web/src/stores/ip.ts
import { defineStore } from 'pinia'
import { client } from '@/api/client'
import { IpService } from '@/api/pb/ip_pb'
import { ConnectError } from '@connectrpc/connect'

const ipService = client(IpService)

export const useIpStore = defineStore('ip', {
  state: () => ({
    loading: false,
    subnetId: 0,
    total: 0,
    items: [],
    id: 0,
    item: {
      id: 0,
      ip: '',
      customerId: 0,
      customerName: '',
      serviceId: 0,
      serviceName: '',
      description: '',
    }
  }),
  actions: {
    setSubnetId(subnetId: number) {
      this.subnetId = subnetId
    },

    setId(id: number) {
      this.id = id
    },

    async getIps() {
      try {
        const { total, items } = await ipService.getIps({
          subnetId: this.subnetId
        })
        this.total = total
        this.items = items?.map((item) => {
          return {
            ...item,
            id: Number(item.id),
            customerId: Number(item.customerId)
          }
        })
      } catch (err) {
        if (err instanceof ConnectError) {
          console.log(err.message)
        }
        throw err
      }
    },

    async getIpById() {
      this.loading = true
      try {
        const res = await ipService.getIpById({ id: this.id })
        this.item = {
          ...res,
          id: Number(res.id),
          serviceId: Number(res.serviceId),
          customerId: Number(res.customerId),
        }
        this.loading = false
      } catch (err) {
        if (err instanceof ConnectError) {
          console.log(err.message)
        }
        this.loading = false
        throw err
      }
    },

    async editCustomerById(id: number, customerId: number) {
      try {
        return await ipService.editIpCustomer({
          id: id,
          customerId: customerId
        })
      } catch (err) {
        if (err instanceof ConnectError) {
          console.log(err.message)
        }
        throw err
      }
    },

    async editServiceById(id: number, serviceId: number) {
      try {
        return await ipService.editIpService({
          id: id,
          serviceId: serviceId
        })
      } catch (err) {
        if (err instanceof ConnectError) {
          console.log(err.message)
        }
        throw err
      }
    },

    async editDescriptionById(id: number, description: string) {
      try {
        return await ipService.editIpDescription({
          id: id,
          description: description
        })
      } catch (err) {
        if (err instanceof ConnectError) {
          console.log(err.message)
        }
        throw err
      }
    }
  },
  getters: {
    getLoading: state => state.loading,

    getTotal: state => state.total,

    getItems: state => state.items,

    getItem: state => state.item,
  }
})
